### PR TITLE
CMake: Various fixes for the cuda configuration

### LINKED ISSUE #3029: Macros undefined when using nvcc
From @kalj in #3005: 
I now found the problem with the DEBUG macro not being defined. Several macro definitions are eventually appended to COMPILE_DEFINITIONS, which are then used internally by CMake to create definitions for a given target. The problem is that for CUDA, where we e.g. use CUDA_ADD_EXECUTABLE instead of ADD_EXECUTABLE, these are not at all picked up. I've mailed the cmake mailing list about this, but I have not had any luck with help from there before. Maybe we have to make yet another hack ourselves...

### LINKED ISSUE #3743: Build system error when using CUDA and cmake 3.7
I can't configure deal with cuda when using cmake 3.7. Everything works fine with cmake 3.6. 
cc: @kalj 

CMake Error at /home/bruno/spack/opt/spack/linux-ubuntu16-x86_64/gcc-6.2.0/cmake-3.7.1-m2qgd6oxyuihv6ildladecgkbdub357x/share/cmake-3.7/Modules/FindCUDA.cmake:1512 (file):
  Error evaluating generator expression:

    $<TARGET_PROPERTY:obj_lac_debug,INCLUDE_DIRECTORIES>

  Target "obj_lac_debug" not found.
Call Stack (most recent call first):
  cmake/macros/macro_deal_ii_add_library.cmake:55 (CUDA_WRAP_SRCS)
  source/lac/CMakeLists.txt:150 (DEAL_II_ADD_LIBRARY)

### PR DESCRIPTION
**Pull request by tamiko**
2bbfcea (Matthias Maier, 86 seconds ago)
   add a changes file

bd87c61 (Matthias Maier, 4 minutes ago)
   CMake: Make sure the cuda wrapper only uses -std=c++11

b84e515 (Matthias Maier, 10 minutes ago)
   CMake: Properly propagate all compiler flags to the cuda wrapper

75eb5b8 (Matthias Maier, 30 minutes ago)
   CMake: Rename all object targets to obj_<foo>_(debug|release)

   This change is necessary to please the cuda wrapper nvcc that insists on
   exporting the target name as preprocessor definition. Unfortunately, a
   period "." is an invalid character and we end up with warnings of the form

     "missing whitespace after the macro name"

   Fix this by renaming all object targets to only contain underscores.

38a1114 (Matthias Maier, 41 minutes ago)
   CMake: Avoid bogus preprocessor definitions

   The macro CUDA_WARP_SRCS expects the target to have a set of preprocessor
   definitions that does not contain an empty element.

48a1f5a (Matthias Maier, 78 minutes ago)
   CMake: Fix a typo

Closes #3743
Closes #3029

### CODE REVIEW

**Comment by tamiko** [comment]
@kalj With this PR the compiler wrapper uses all host compiler flags and definitions.

File: cmake/macros/macro_deal_ii_add_library.cmake [hunk 1]
```diff
@@ -37,9 +37,19 @@ MACRO(DEAL_II_ADD_LIBRARY _library)
...
+    # elements lightly...
+    #
+    SET(_definitions
+      ${DEAL_II_DEFINITIONS}
+      ${DEAL_II_DEFINITIONS_${_build}}
+      )
+
     SET_TARGET_PROPERTIES(${_library}.${_build_lowercase} PROPERTIES
       LINK_FLAGS "${DEAL_II_LINKER_FLAGS} ${DEAL_II_LINKER_FLAGS_${_build}}"
-      COMPILE_DEFINITIONS "${DEAL_II_DEFINITIONS};${DEAL_II_DEFINITIONS_${_build}}"
```
Comment (kalj) [inline cmake/macros/macro_deal_ii_add_library.cmake]: Care to elaborate a bit on what is going on here?

**Comment by kalj** [comment]
@tamiko This looks great, but if I understand correctly, the workaround is to add all the definitions *before* calling `CUDA_WRAP_SRCS`. If that is the case, and we still cannot add stuff afterwards, it seems hard to make `DEAL_II_SETUP_TARGET` do things correctly for a user application. Do you agree? Would that require the previously discussed customization of FindCUDA itself?

**Comment by kalj** [comment]
Looks like you missed one place: https://github.com/tamiko/dealii/pull/7

**Comment by tamiko** [comment]
@kalj - Thanks, I applied your commit.

@kalj 
> This looks great, but if I understand correctly, the workaround is to add all the definitions before calling CUDA_WRAP_SRCS. If that is the case, and we still cannot add stuff afterwards, it seems hard to make DEAL_II_SETUP_TARGET do things correctly for a user application. Do you agree?

I do not follow. No, I revise my opinion - I think the `CUDA_WRAP_SRCS` macro is sufficient and stable enough for our purposes. We will have to check future CMake releases for surprises, though.

**Comment by tamiko** [comment]
@kalj The `CUDA_WRAP_SRCS` macro uses
 - the specified target to query for compile definitions
 - `CMAKE_CXX_FLAGS` for compiler flags

So, by providing a valid target, we fix the compile definition issue. Further, by populating `CMAKE_CXX_FLAGS` prior to compilation we work around the fact that `CUDA_WRAP_SRCS` does not query the target for compile flags. The latter is a workaround.

**Comment by kalj** [comment]
@tamiko What I'm trying to say is that, since `CUDA_WRAP_SRCS` doesn't work with `SET_PROPERTY` (https://gitlab.kitware.com/cmake/cmake/issues/16401), we cannot let the user do stuff like:

    CUDA_ADD_EXECUTABLE(my_application my_application.cu)
    DEAL_II_SETUP_TARGET(my_application)

like we can do for non-CUDA code:

    ADD_EXECUTABLE(my_application my_application.cc)
    DEAL_II_SETUP_TARGET(my_application)

Of course, if we only support using `DEAL_II_INVOKE_AUTOPILOT`, then we can do anything before calling `CUDA_ADD_EXECUTABLE`, which probably lets us set things up equivalently, but in my opinion, that would be a bit restrictive. For instance, I like the way that `SET_PROPERTY` does not pollute the general `CMAKE_CXX_FLAGS`, so that different targets in your project can have different configurations. What is your take on this?

**Comment by kalj** [comment]
More specifically, do you think the same flexibility and functionality could be achieved as is, or using some other way, without digging into `CUDA_WRAP_SRCS`?

**Comment by tamiko** [comment]
>     CUDA_ADD_EXECUTABLE(my_application my_application.cu)
>     DEAL_II_SETUP_TARGET(my_application)

No, that doesn't work.

There are multiple ways to achieve this behavior, e.g. out of the box:

 [ ... set compile definitions and cxx flags globally ...]
 CUDA_ADD_EXECUTABLE(target ...)
 DEAL_II_SETUP_TARGET(target)

Or, by providing a macro that simply sets up the cuda files:

  ADD_LIBRARY(obj_target OBJECT ...)
  DEAL_II_SETUP_TARGET(obj_target)
  DEAL_II_SETUP_CUDA(obj_target obj_target_cuda_files)

  ADD_EXECUTABLE(target
    $<TARGET_OBJECTS:obj_target>
    ${obj_target_cuda_files}
    )

Alternatively, and maybe preferred:

  DEAL_II_ADD_CUDA_EXECUTABLE(target ...)

**Comment by Rombur** [comment]
LGTM What do you think @kalj ?

**Comment by kalj** [comment]
@Rombur Yes, definitely! There is still some more work to do on the CUDA build system, but this is a great step forward.

**Comment by kalj** [comment]
@tamiko I think I would go with `DEAL_II_CUDA_ADD_EXECUTABLE`, but maybe even use a general `DEAL_II_ADD_EXECUTABLE` and keep the CUDA-specific stuff hidden. I still think we should have a long-term goal of fixing `CUDA_WRAP_SRCS` to allow for something like what I suggested. It would be good to use upstream stuff as much as possible, and avoid intrusive custom macros.

**Pull request merged**